base/dates/periods.jl

```diff
@@ -166,10 +166,44 @@ type CompoundPeriod <: AbstractTime
                     break
                 end
             end
-            return new(reverse!(pc))
+            p = reverse(pc)
+            n = length(p)
         else
             return new(resize!(p, n))
         end
+        # canonicalize Period values such that they are all either
+        # positive or negative.
```

Comment:
@stevengj code should now be working as expected. I also updated the comment to reflect the new behaviour.